projects/packages/connection/src/class-nonce-handler.php

```diff
@@ -0,0 +1,202 @@
+ */
+
+namespace Automattic\Jetpack\Connection;
+
+/**
+ * The nonce handler.
+ */
+class Nonce_Handler {
+
+	/**
+	 * How long the scheduled cleanup can run (in seconds).
+	 * Can be modified using the filter `jetpack_connection_nonce_scheduled_cleanup_limit`.
+	 */
+	const SCHEDULED_CLEANUP_TIME_LIMIT = 5;
+
+	/**
+	 * How many nonces should be removed per batch during the `clean_all()` run.
+	 */
+	const CLEAN_ALL_LIMIT_PER_BATCH = 1000;
+
+	/**
+	 * Nonce lifetime in seconds.
+	 */
+	const LIFETIME = HOUR_IN_SECONDS;
+
+	/**
+	 * The nonces used during the request are stored here to keep them valid.
+	 * The property is static to keep the nonces accessible between the `Nonce_Handler` instances.
+	 *
+	 * @var array
+	 */
+	private static $nonces_used_this_request = array();
+
+	/**
+	 * The database object.
+	 *
+	 * @var \wpdb
+	 */
+	private $db;
+
+	/**
+	 * Initializing the object.
+	 */
+	public function __construct() {
+		global $wpdb;
+
+		$this->db = $wpdb;
+	}
+
+	/**
+	 * Scheduling the WP-cron cleanup event.
+	 */
+	public function init_schedule() {
+		add_action( 'jetpack_clean_nonces', array( __CLASS__, 'clean_scheduled' ) );
+		if ( ! wp_next_scheduled( 'jetpack_clean_nonces' ) ) {
+			wp_schedule_event( time(), 'hourly', 'jetpack_clean_nonces' );
+		}
+	}
+
+	/**
+	 * Reschedule the WP-cron cleanup event to make it start sooner.
+	 */
+	public function reschedule() {
+		wp_clear_scheduled_hook( 'jetpack_clean_nonces' );
+		wp_schedule_event( time(), 'hourly', 'jetpack_clean_nonces' );
+	}
+
+	/**
+	 * Adds a used nonce to a list of known nonces.
+	 *
+	 * @param int    $timestamp the current request timestamp.
+	 * @param string $nonce the nonce value.
+	 *
+	 * @return bool whether the nonce is unique or not.
+	 */
+	public function add( $timestamp, $nonce ) {
+		if ( isset( static::$nonces_used_this_request[ "$timestamp:$nonce" ] ) ) {
+			return static::$nonces_used_this_request[ "$timestamp:$nonce" ];
+		}
+
+		// This should always have gone through Jetpack_Signature::sign_request() first to check $timestamp and $nonce.
+		$timestamp = (int) $timestamp;
+		$nonce     = esc_sql( $nonce );
+
+		// Raw query so we can avoid races: add_option will also update.
+		$show_errors = $this->db->hide_errors();
+
+		// Running try...finally to make sure.
+		try {
+			$old_nonce = $this->db->get_row(
+				$this->db->prepare( "SELECT 1 FROM `{$this->db->options}` WHERE option_name = %s", "jetpack_nonce_{$timestamp}_{$nonce}" )
+			);
+
+			if ( is_null( $old_nonce ) ) {
+				$return = (bool) $this->db->query(
+					$this->db->prepare(
+						"INSERT INTO `{$this->db->options}` (`option_name`, `option_value`, `autoload`) VALUES (%s, %s, %s)",
+						"jetpack_nonce_{$timestamp}_{$nonce}",
+						time(),
+						'no'
+					)
+				);
+			} else {
+				$return = false;
+			}
+		} finally {
+			$this->db->show_errors( $show_errors );
+		}
+
+		static::$nonces_used_this_request[ "$timestamp:$nonce" ] = $return;
```

Comment:
Yes, exactly.
If the nonce is not in the `$nonces_used_this_request` array, but already in the database, that means this is a repeated request with the same nonce, which should be blocked.
In that case the `false` gets stored in the array as a cache, to prevent the repeated database queries from being run.